uroot_test.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2015-2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/u-root/u-root/pkg/cpio"
+	"github.com/u-root/u-root/pkg/testutil"
+	itest "github.com/u-root/u-root/pkg/uroot/initramfs/test"
+)
+
+var twocmds = []string{
+	"github.com/u-root/u-root/cmds/ls",
+	"github.com/u-root/u-root/cmds/init",
+}
+
+var srcmds = []string{
+	"github.com/u-root/u-root/cmds/ls",
+	"github.com/u-root/u-root/cmds/init",
+	"github.com/u-root/u-root/cmds/installcommand",
+}
+
+type buildSourceValidator struct {
+	gopath string
+	goroot string
+	env    []string
+}
+
+func (b buildSourceValidator) Validate(a *cpio.Archive) error {
+	dir, err := ioutil.TempDir("", "u-root-source-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// Unpack into dir.
+	err = cpio.ForEachRecord(a.Reader(), func(r cpio.Record) error {
+		return cpio.CreateFileInRoot(r, dir, false)
+	})
+	if err != nil {
+		return err
+	}
+
+	goroot := filepath.Join(dir, b.goroot)
+	gopath := filepath.Join(dir, b.gopath)
+	// go build ./src/...
+	c := exec.Command(filepath.Join(goroot, "bin/go"), "build", filepath.Join(gopath, "src/..."))
+	c.Env = append(b.env,
+		fmt.Sprintf("GOPATH=%s", gopath),
+		fmt.Sprintf("GOROOT=%s", goroot))
+	out, err := c.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not build go source %v; output\n%s", err, out)
+	}
+	return nil
+}
+
+func TestUrootCmdline(t *testing.T) {
+	t.Parallel()
```

Comment:
This doesn't make the TestUrootCmdline test any faster. See https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks